File: figaro-nats/src/figaro_nats/streams.py
"""JetStream stream configuration for Figaro."""

from __future__ import annotations

import asyncio
import logging

from nats.js import JetStreamContext
from nats.js.api import StreamConfig, RetentionPolicy

logger = logging.getLogger(__name__)

TASKS_STREAM = "TASKS"
TASKS_SUBJECTS = ["figaro.task.>"]
TASKS_MAX_AGE = 7 * 24 * 60 * 60  # 7 days in seconds

HELP_STREAM = "HELP"
HELP_SUBJECTS = ["figaro.help.*.response"]
HELP_MAX_AGE = 24 * 60 * 60  # 1 day in seconds


async def _ensure_stream(
    js: JetStreamContext,
    name: str,
    subjects: list[str],
    max_age: int,
) -> None:
    """Create or update a single JetStream stream."""
    config = StreamConfig(
        name=name,
        subjects=subjects,
        retention=RetentionPolicy.LIMITS,
        max_age=max_age,
    )
    try:
        await js.find_stream_name_by_subject(subjects[0])
        await js.update_stream(config)
        logger.info("Updated JetStream stream: %s", name)
    except Exception:
        await js.add_stream(config)
        logger.info("Created JetStream stream: %s", name)


async def ensure_streams(js: JetStreamContext) -> None:
    """Create or update JetStream streams required by Figaro."""
    await asyncio.gather(
        _ensure_stream(js, TASKS_STREAM, TASKS_SUBJECTS, TASKS_MAX_AGE),
        _ensure_stream(js, HELP_STREAM, HELP_SUBJECTS, HELP_MAX_AGE),
    )

File: figaro-worker/src/nats/streams.ts
/**
 * JetStream stream configuration for Figaro.
 *
 * Ported from figaro-nats/src/figaro_nats/streams.py
 */

import { RetentionPolicy, type NatsConnection } from "nats";

export const TASKS_STREAM = "TASKS";
export const TASKS_SUBJECTS = ["figaro.task.>"];
export const TASKS_MAX_AGE_SECONDS = 7 * 24 * 60 * 60; // 7 days

export const HELP_STREAM = "HELP";
export const HELP_SUBJECTS = ["figaro.help.*.response"];
export const HELP_MAX_AGE_SECONDS = 24 * 60 * 60; // 1 day

/** Convert seconds to nanoseconds (nats.js uses nanos for max_age). */
function secondsToNanos(seconds: number): number {
  return seconds * 1_000_000_000;
}

/**
 * Create or update a single JetStream stream.
 */
async function ensureStream(
  nc: NatsConnection,
  name: string,
  subjects: string[],
  maxAgeSeconds: number,
): Promise<void> {
  const jsm = await nc.jetstreamManager();

  try {
    // Try to get the existing stream and update it
    await jsm.streams.info(name);
    await jsm.streams.update(name, {
      subjects,
      max_age: secondsToNanos(maxAgeSeconds),
    });
    console.log(`[streams] Updated JetStream stream: ${name}`);
  } catch {
    // Stream does not exist yet -- create it
    await jsm.streams.add({
      name,
      subjects,
      retention: RetentionPolicy.Limits,
      max_age: secondsToNanos(maxAgeSeconds),
    });
    console.log(`[streams] Created JetStream stream: ${name}`);
  }
}

/**
 * Create or update all JetStream streams required by Figaro.
 */
export async function ensureStreams(nc: NatsConnection): Promise<void> {
  await Promise.all([
    ensureStream(nc, TASKS_STREAM, TASKS_SUBJECTS, TASKS_MAX_AGE_SECONDS),
    ensureStream(nc, HELP_STREAM, HELP_SUBJECTS, HELP_MAX_AGE_SECONDS),
  ]);
}
